### PR TITLE
Remove unsound "delay (force x) => x" optimisation

### PR DESCRIPTION
**Pull request by L-as**
Thanks @TotallyNotChase for noticing the issue in the generated code!

### CODE REVIEW

**Comment by L-as** [comment]
I'm still debating internally between view patterns and record syntax for pattern matching:
```haskell
TermResult { getTerm = RError }
```
vs
```haskell
getTerm -> RError
```

File: Plutarch/Internal.hs [hunk 1]
```diff
@@ -60,6 +60,20 @@ hashRawTerm' (RHoisted (HoistedTerm hash _)) = flip hashUpdate ("8" :: BS.ByteSt
...
+  }
+
+mapTerm :: (RawTerm -> RawTerm) -> TermResult -> TermResult
+mapTerm f (TermResult t d) = TermResult (f t) d
+
+mergeTerms :: (RawTerm -> RawTerm -> RawTerm) -> TermResult -> TermResult -> TermResult
+mergeTerms f x y = TermResult (f (getTerm x) (getTerm y)) (getDeps x <> getDeps y)
+
+mkNew :: RawTerm -> TermResult
+mkNew r = TermResult r []
```
Comment (TotallyNotChase) [inline Plutarch/Internal.hs:75]: Name this `mkTermRes` maybe? I understand that this isn't exported so it hardly matters but still.

File: Plutarch/Internal.hs [hunk 2]
```diff
@@ -82,64 +96,61 @@ data PDelayed (a :: k -> Type) (s :: k)
 
 plam' :: (Term s a -> Term s b) -> Term s (a :--> b)
 plam' f = Term $ \i ->
-  let v = Term $ \j -> (RVar (j - (i + 1)), [])
-      (t, deps) = asRawTerm (f v) (i + 1)
-   in (RLamAbs $ t, deps)
+  let v = Term $ \j -> TermResult (RVar (j - (i + 1))) []
```
Comment (TotallyNotChase) [inline Plutarch/Internal.hs:99]:
```hs
TermResult (RVar (j - (i + 1))) []
```
Is there a reason for not using `mkNew` here?

**Review by TotallyNotChase — APPROVED** [review]

**Comment by TotallyNotChase** [comment]
I tested this on the validator rewrite and it works correctly! Looks like all the optimizations combined also shaved off some script size and cpu/mem units!

File: Plutarch/Internal.hs (hunk 2 above)
Comment (L-as) [reply Plutarch/Internal.hs:99]: Nice catch.

File: Plutarch/Internal.hs (hunk 1 above)
Comment (L-as) [reply Plutarch/Internal.hs:75]: Thanks.

**Review by TotallyNotChase — APPROVED** [review]

**Comment by TotallyNotChase** [comment]
LGTM! Let's get this onto production!

**Pull request merged**